File: web/src/lib/play/dialogueSegments.ts
export interface ProseSegment {
  type: 'narrative' | 'dialogue';
  text: string;
}

const QUOTE_PAIRS: Array<[string, string]> = [
  ['«', '»'],
  ['"', '"'],
  ['“', '”'],
];

const OPENERS = new Map<string, string>(QUOTE_PAIRS);

/** Lines starting with one of these dashes read as spoken dialogue. */
const DASH_LINE = /^\s*(—|–|-)\s+/;

/**
 * Splits scene prose into narrative vs dialogue segments.
 * Dialogue = text inside quotation marks («», "", "", “”) or dash-led lines.
 * An unterminated opening quote styles the trailing text as dialogue.
 * Adjacent same-type segments are merged.
 */
export function segmentProse(input: string): ProseSegment[] {
  const out: ProseSegment[] = [];
  const push = (type: ProseSegment['type'], text: string) => {
    if (!text || !text.trim()) return;
    const last = out[out.length - 1];
    if (last && last.type === type) {
      last.text += '\n' + text;
    } else {
      out.push({ type, text });
    }
  };

  const lines = (input || '').split('\n');
  for (const line of lines) {
    if (DASH_LINE.test(line)) {
      push('dialogue', line.replace(DASH_LINE, '').trim());
      continue;
    }
    let i = 0;
    let buf = '';
    while (i < line.length) {
      const ch = line[i];
      const closer = OPENERS.get(ch);
      if (closer) {
        const end = line.indexOf(closer, i + 1);
        if (buf) {
          push('narrative', buf);
          buf = '';
        }
        if (end === -1) {
          push('dialogue', line.slice(i + 1).trim());
          i = line.length;
        } else {
          push('dialogue', line.slice(i + 1, end).trim());
          i = end + 1;
        }
      } else {
        buf += ch;
        i += 1;
      }
    }
    if (buf) push('narrative', buf);
  }
  return out.map((s) => ({ ...s, text: s.text.trim() })).filter((s) => s.text.length > 0);
}

File: app/lib/core/utils/dialogue_segments.dart
// Splits scene prose into narrative vs dialogue segments.
// Dialogue = text inside quotation marks («», "", “”) or dash-led lines.
// Mirrors web `src/lib/play/dialogueSegments.ts` — keep the two in sync.

enum ProseSegmentType { narrative, dialogue }

class ProseSegment {
  final ProseSegmentType type;
  String text;
  ProseSegment(this.type, this.text);
}

const _closers = <String, String>{
  '«': '»',
  '"': '"',
  '“': '”',
};

final _dashLine = RegExp(r'^\s*(—|–|-)\s+');

List<ProseSegment> segmentProse(String input) {
  final out = <ProseSegment>[];
  void push(ProseSegmentType type, String text) {
    if (text.trim().isEmpty) return;
    if (out.isNotEmpty && out.last.type == type) {
      out.last.text += '\n$text';
    } else {
      out.add(ProseSegment(type, text));
    }
  }

  for (final line in input.split('\n')) {
    final dash = _dashLine.firstMatch(line);
    if (dash != null) {
      push(ProseSegmentType.dialogue, line.substring(dash.end).trim());
      continue;
    }
    var i = 0;
    final buf = StringBuffer();
    void flush() {
      if (buf.isNotEmpty) {
        push(ProseSegmentType.narrative, buf.toString());
        buf.clear();
      }
    }

    while (i < line.length) {
      final ch = line[i];
      final closer = _closers[ch];
      if (closer != null) {
        final end = line.indexOf(closer, i + 1);
        flush();
        if (end == -1) {
          push(ProseSegmentType.dialogue, line.substring(i + 1).trim());
          i = line.length;
        } else {
          push(ProseSegmentType.dialogue, line.substring(i + 1, end).trim());
          i = end + 1;
        }
      } else {
        buf.write(ch);
        i += 1;
      }
    }
    flush();
  }
  return out
      .map((s) => ProseSegment(s.type, s.text.trim()))
      .where((s) => s.text.isNotEmpty)
      .toList();
}
